utils/DMTHandler.js

```diff
@@ -1,154 +1,193 @@
 /* eslint-disable prefer-promise-reject-errors */
 /* eslint-disable consistent-return */
-import { replacer } from './utils';
+import { replacer, removeTrailingSlashes } from './utils';
 import logger from './logUtil';
 
-// default values for retry
-const retryCount = 3;
+class DMTHandler {
+  constructor(writeKey, dataPlaneUrl) {
+    this.dataPlaneUrl = removeTrailingSlashes(dataPlaneUrl);
+    this.writeKey = writeKey;
+    this.retryCount = 3; // default value for retry
+    this.queue = [];
+    this.isTransformationProcessing = false;
+    this.isTimerStarted = false;
+  }
 
-/**
- * A helper function that will take rudderEelement as an event and generate
- * a batch payload that will be sent to transformation server
- *
- */
-const createPayload = (event) => {
-  const orderNo = Date.now();
-  const payload = {
-    batch: [
-      {
-        orderNo,
-        event,
-      },
-    ],
-  };
-
-  return payload;
-};
+  // Enqueue the events and callbacks
+  enqueue(event, cb) {
+    this.queue.push({
+      event,
+      cb,
+    });
+    this.start();
+  }
 
-/**
- * A helper function that will take one payload at a time and send it to transformation server
- * and return the response
- *
- */
-const sendEventForTransformation = (payload, writeKey, dataPlaneUrl, retryCount) => {
-  return new Promise((resolve, reject) => {
-    // const url = `${dataPlaneUrl}/transform`;
-    const url = 'https://shadowfax-dataplane.dev-rudder.rudderlabs.com/transform';
-    const headers = {
-      'Content-Type': 'application/json',
-      Authorization: `Basic ${btoa(`${writeKey}:`)}`,
+  /**
+   * A helper function that will take rudderEelement as an event and generate
+   * a batch payload that will be sent to transformation server
+   *
+   */
+  createPayload(event) {
+    const orderNo = Date.now();
+    const payload = {
+      batch: [
+        {
+          orderNo,
+          event,
+        },
+      ],
     };
-    try {
-      const xhr = new XMLHttpRequest();
-      xhr.open('POST', url, true);
-      Object.keys(headers).forEach((k) => xhr.setRequestHeader(k, headers[k]));
 
-      xhr.onreadystatechange = () => {
-        if (xhr.readyState === 4) {
-          try {
-            const { status } = xhr;
-            if (status === 200) {
-              let { response } = xhr;
-              if (response && typeof response === 'string') {
-                response = JSON.parse(response);
-              } else {
-                reject(`[Transformation]:: Transformation failed. Invalid response from server.`);
+    return payload;
+  }
+
+  /**
+   * A helper function that will take one payload at a time and send it to transformation server
+   * and return the response
+   *
+   */
+  sendEventForTransformation(payload, retryCount) {
+    return new Promise((resolve, reject) => {
+      // const url = `${this.dataPlaneUrl}/transform`;
+      const url = 'https://shadowfax-dataplane.dev-rudder.rudderlabs.com/transform';
```

Comment:
This should never end up in the commits pushed to the repository.

Reply:
Removed.